grpc/server/src/test/java/io/helidon/grpc/server/TracingIT.java

```diff
@@ -0,0 +1,149 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.grpc.server;
+
+import java.net.URI;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.LogManager;
+import java.util.logging.Logger;
+
+import io.helidon.grpc.server.test.Echo;
+import io.helidon.grpc.server.test.EchoServiceGrpc;
+import io.helidon.tracing.TracerBuilder;
+
+import io.grpc.Channel;
+import io.grpc.ManagedChannelBuilder;
+import io.opentracing.Tracer;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import services.EchoService;
+import zipkin2.Span;
+import zipkin2.junit.ZipkinRule;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Tests for gRPC server with tracing.
+ *
+ * @author Bin Chen
+ */
+public class TracingIT {
+
+    // ----- data members ---------------------------------------------------
+
+    /**
+     * The {@link java.util.logging.Logger} to use for logging.
+     */
+    private static final Logger LOGGER = Logger.getLogger(TracingIT.class.getName());
+
+    /**
+     * The Helidon {@link io.helidon.grpc.server.GrpcServer} being tested.
+     */
+    private static GrpcServer grpcServer;
+
+    /**
+     * A gRPC {@link io.grpc.Channel} to connect to the test gRPC server
+     */
+    private static Channel channel;
+
+    /**
+     * ZipkinRule to start a Zipkin server
+     */
+    public static ZipkinRule zipkin = new ZipkinRule();
+
+
+    // ----- test lifecycle -------------------------------------------------
+
+    @BeforeAll
+    public static void setup() throws Exception {
+        LogManager.getLogManager().readConfiguration(TracingIT.class.getResourceAsStream("/logging.properties"));
+
+        //start zipkin server on an ephemeral port
+        zipkin.start(0);
+
+        startGrpcServer();
+
+        channel = ManagedChannelBuilder.forAddress("localhost", grpcServer.port())
+                                       .usePlaintext()
+                                       .build();
+    }
+
+    @AfterAll
+    public static void cleanup() throws Exception {
+        zipkin.shutdown();
+        }
+
+    // ----- test methods ---------------------------------------------------
+
+    @Test
+    public void shouldTraceMethodNameAndHeaders() throws Exception{
+        // call the gRPC Echo service so that there should be tracing span sent to zipkin server
+        EchoServiceGrpc.newBlockingStub(channel).echo(Echo.EchoRequest.newBuilder().setMessage("foo").build());
+
+        // wait a bit for span from zipkin
+        Thread.sleep(2000);
```

Comment:
We can probably introduce Bedrock end use `Eventually.assertThat()` instead of sleeping. We'll have to see then whether the Helidon team object to us using Bedrock. You will need to add a dependency on groupId `com.oracle.bedrock` artifactId `bedrock-testing-support` version `5.0.11` and put it into the top level pom.xml and then into this project's pom.xml without the version element.

As Bedrock creates a proxy (mock) of the class used in `Eventually.assertThat()` we cannot use the `zipkin` value directly as the `ZipkinRule` class is final so it cannot be proxied. So, you need to do this in a bit of a round-about way. First create a method in this class like this:
```
public int getSpanCount() {
    return zipkin.collectorMetrics().spans();
}
```
Then instead of the sleep you can do:
```
Eventually.assertThat(invoking(this).getSpanCount(), is(1));
```
Now inside `invoking` we are creating a proxy of this test class, which will work fine. This will then assert that eventually there is at least one span then proceed with the rest of your test.

Reply:
Yeah, we definitely shouldn't sleep... We know that random sleeps caused all kinds of issues in Coherence tests and that we went through hell to get rid of them, so we shouldn't even start with them here.

I think Bedrock is fine, and Helidon guys won't object to it, especially if the WLS team is starting to use it as well, but we should definitely discuss it with them next time we meet.